### ISSUE TITLE
Proposal: Test Aware Container (making Docker "test-friendly")

### ISSUE THREAD

**Issue by alexei-led** [docker/docker#21280]
## Test Aware Container

Today, Docker container lifecycle is naively focused on:

 > Build -> Ship -> Run

 Build the image, ship it to some registry, and run it anywhere. IMHO there's a critical missing step - **Test**. 

I would like to propose to make Docker engine **"test friendly"**. There should be a standard way not only to build, deploy and run Docker container, but also to test it.

The more complete sequence should be, like this:

> Build -> **Test** -> Ship -> Run

People are doing this today, by embedding testing tools, tests, required packages and scripts into application container. This pollutes the application image with unneeded binaries and test related artefacts. Of cause, it is possible to keep testing outside of container, on you CI server, for example, thus getting to the situation "these tests pass in CI, I have no idea why they fail on your environment..."

Here, I would like to propose and extension to Dockerfile syntax to make it "test-friendly" and also add a new Docker command. 

```
ONTEST [INSTRUCTION]
```

The new `ONTEST` instruction is essentially similar to existing [`UNBUILD`](https://docs.docker.com/engine/reference/builder/#onbuild) instruction. The `ONTEST` instruction adds a trigger instruction to the image to be executed at a later time when the image is tested. Any build instruction can be registered as a trigger.

The `ONTEST` instruction should be recognized by a new `docker test` command.

```
docker test [OPTIONS] IMAGE [COMMAND] [ARG...]
```

The `docker test` command syntax should be similar to `docker run` command, with one significant difference: a new "testable" image could be automatically generated and even tagged with `<image name>:<image tag>-test` tag ("test" postfix added to the original image tag). This "testable" image will be generated `FROM` the application image, executing all build instructions, defined after `ONTEST` command and executing `ONTEST CMD` (or `ONTEST ENTRYPOINT`).
The `docker test` command should return a non-zero code if any tests fail. The test results should be written into an automatically generated `VOLUME` that points to `/var/tests/results` folder.

The proposed approach allows not only to separate between application and tests, it also creates a special image that contains both application and tests (including testing tool). This "testing" image allows to achieve test portability. Test portability allows you to run same testing tools, same tests, same environment on any environment: CI, developer, staging, production, etc. 

Please take a look at sample `Dockerfile` below - it includes the new proposed `ONTEST` instruction.

```dockerfile
FROM "<base image>":"<version>"

WORKDIR "<path>"

# install packages required to run app
RUN apt-get update && apt-get install -y \
    "<app runtime> and <dependencies>" \  # add app runtime and required packages
    && rm -rf /var/lib/apt/lists/*

# install packages required to run tests   
ONTEST RUN apt-get update && apt-get install -y \
           "<test tools> and <dependencies>"    \     # add testing tools and required packages
           && rm -rf /var/lib/apt/lists/*

# copy app files
COPY app app
COPY run.sh run.sh

# copy test scripts
ONTEST COPY tests tests

# copy "main" test command
ONTEST COPY test.sh test.sh

# auto-generated volume for test results
# ONTEST VOLUME "/var/tests/results"

# ... EXPOSE, RUN, ADD ... for app and test environment

# main app command
CMD [run.sh, "<app arguments>"]

# main test command
ONTEST CMD [/test.sh, "<test arguments>"]
```
![test_aware](https://cloud.githubusercontent.com/assets/1898375/13842749/97bf11f4-ec38-11e5-9265-34924ed617b0.png)

**Note**: The original idea and other container testing approaches are described in this [blog post](https://medium.com/@alexeiled/testing-strategies-for-docker-containers-f633e261e75a#.y097ymjzr)

**Comment by phemmer** [comment]
You already can do this by just using 2 `Dockerfile`s.

```
FROM "<base image>":"<version>"

WORKDIR "<path>"

# install packages required to run app
RUN apt-get update && apt-get install -y \
    "<app runtime> and <dependencies>" \  # add app runtime and required packages
    && rm -rf /var/lib/apt/lists/*

# copy app files
COPY app app
COPY run.sh run.sh

# main app command
CMD [run.sh, "<app arguments>"]
```

```
FROM "<myimage>:<version>"

# install packages required to run tests   
ONTEST RUN apt-get update && apt-get install -y \
           "<test tools> and <dependencies>"    \     # add testing tools and required packages
           && rm -rf /var/lib/apt/lists/*

# copy test scripts
ONTEST COPY tests tests

# copy "main" test command
ONTEST COPY test.sh test.sh

# auto-generated volume for test results
# ONTEST VOLUME "/var/tests/results"

# ... EXPOSE, RUN, ADD ... for app and test environment
```

This also seems like a highly opinionated stance on how tests should be done. The current docker model offers a fair amount of flexibility by letting you use your containers how you want. Implementing this would be docker saying "hey, you should be testing like this", and it may not be what everyone needs.

> The docker test command should return a non-zero code if any tests fail. The test results should be written into an automatically generated VOLUME that points to /var/tests/results folder.

What if I want my tests results to go to STDOUT? I don't want to have to use `docker cp` to get to them. Unless you're saying it's only generated files which go there. In which case what if I can't control where the test utilities put their files (maybe they get dropped in the root of the project)?


Also, in order to support this, the builder will have to take all the `ONTEST` statements in the Dockerfile, and sort them to the end (after all the non-ONTEST statements). I think this would add a lot of confusion from people.

**Comment by alexei-led** [comment]
The same statement is true for ONBUILD instruction - it also can be replaced with 2 separately managed Dockerfiles. This is not the point.
The idea is to bring some common convention around testing of containerized application. This will simplify testing process and make it easier to create a supporting tools. For example, any CI tool will be able to test app and collect test results, without a need to configure it for every single project.

**Comment by AndrewGuenther** [comment]
Why can't you have a test image which has your production image as a parent? That's what I've always done and it works just fine.

**Comment by alexei-led** [comment]
@AndrewGuenther thanks for comment. That's what we do too: making test image `FROM` app image.
The proposal is about defining a convention ([convention over confguration](https://en.wikipedia.org/wiki/Convention_over_configuration) design principle). 
Once there is a common **testing** convention, things can be simplified: CI automation, test execution, test result analytics and etc.
Like with any API or protocol, once you define it, a lot of tools and integrations appear around it. This is the idea: 
> Make testing and quality automation - the first class citizen in the Docker container world.

**Comment by dnephin** [comment]
I think the problem with this proposal is it assumes there is only one kind of test suite. In reality there are many different types of tests, all of which can already exist without introducing new commands and Dockerfile instructions.

End-to-end tests of an image should be performed from outside the image, treating the image as a black box. No installation of extra software or tools inside the image is required.

Unit or integration testing can be done using a separate image. It's ok that this image is different from the production image because 1) you're only testing components of the final product 2) you have end-to-end tests to cover the image itself.

> The idea is to bring some common convention around testing of containerized application

I like this idea, but I think the Dockerfile is the wrong place for it. These conventions should live in higher level systems not at the image/Dockerfile laayer.

**Comment by AndrewGuenther** [comment]
My biggest problem with the proposal as it currently stands is that I think of it like including unit tests in your source files.

I completely agree with @dnephin. Why not just a best practices documentation page?

**Comment by alexei-led** [comment]
@dnephin I agree with you, that end-to-end tests should be performed outside of the application image and in this case testing tools and scripts should NOT be packaged together with application code into the same image.

What I'm suggesting, is to package testing tools and scripts also into "some" Docker image (it can be same as app image, or separate image). By packaging testing tools and scripts into special "test" image, you also is able to "capture" testing, at specific point of time (image build) and specific configuration and test scripts. This also makes testing portable, like app image. You will be able to run same tests with same testing tools and config on any machine (CI, dev, production, etc). Isn't it nice? The way you will run tests should be the same, no matter if container contains only tests or tests and app.

Since, Docker defines the way of how to build and run app code, i.e. it sets its POV on Application Development Lifecycle, I think, its worth also to specify how test containers: how internal or external testing should be done. Testing is important stage in software development and having some kind of standard will allow to grow an eco system of tools and integration around Docker quality.

**Comment by philipz** [comment]
@alexei-led Just use Docker Compose to build up ALL enviroment to do integration test, like [this](https://circleci.com/gh/philipz/circlecitest/2).
![docker_compose_test](https://cloud.githubusercontent.com/assets/664465/17726719/5bdea074-6487-11e6-85ae-5eb5ae4d812f.png)

**Comment by alexei-led** [comment]
@philipz your example with compose is typical CI setup, but it's totally orthogonal to the proposed idea. Your tests are not portable - you need to do initial setup on every machine (CI, dev, lab) where you want to run these tests and make sure that this setup is "immutable" (like container): test scripts, runtime (node version and npm modules), environment variables, host packages, etc. Currently you can run your tests on CI machine only, your dev machine have similar environment (maybe), but I, for example, cannot run your tests (I do not have node installed on my laptop).
So, making "Test Containers" allows to achieve portability also for automated tests. Exactly same tests (and test tools) can be run on ANY machine, without ANY setup - only Docker engine is required. 
Once, there is a "standard" way to create, run and get test results from Test Containers, it will be easy to take testing to the same level as `docker build`, where everyone knows how to build container, but there is no standard way to test containers and collect test results (or even decide if test passed or failed).

**Comment by philipz** [comment]
@alexei-led yes, my case was use Node.js environment right, but it is portable absolutely if ALL environments are dockerization(DEV, STAGE, PRD). Just use test container of Node.js to run test scripts can fix the problem. And your blog post is great, but ADD test scripts into image is bad idea. Test program and main program should be into different project and build different docker image.

**Comment by alexei-led** [comment]
There are several options to package tests into Test Containers. If these are integration tests, its better to create a separate (from app) container with test scripts and test engine. Some tests require access to app code; for these you need to package them together with app image.

**Comment by whomper** [comment]
Embedding peripheral functions, like testing or monitoring, into the container image has the effect of producing a self contained entity that could have those functions invoked on it in every step of the pipeline, including production. As more tools will use that concept, so will the current pipeline engines become distributed and much more flexible.

[http://techbeacon.com/optimize-devops-self-contained-containers](url)

**Comment by mitar** [comment]
I also think having `ONTEST` would be great. It would not solve all testing needs, but it would allow for basic things to make sure that the image on its own works as expected. Especially for example when one has some extra patches on top of base Linux distribution, it is great to be able to check if those patches still work as expected, or maybe upstream fixed it or something.

**Comment by jamrizzi** [comment]
I would definitely use this feature

**Comment by krailler** [comment]
I like this.

**Comment by tbflw** [comment]
We're supportive of making Docker more test friendly and support the complete Build -> Test -> Ship -> Run cycle. In our case what we need in our build process, is essentially just multistage builds together with the possibility of mounting a volume into intermediate containers during build in order to get results out from the build steps as described in  #14080.

**Comment by dmsteinberg** [comment]
This is a fantastic idea, except for this part: "The test results should be written into an automatically generated VOLUME that points to /var/tests/results folder."

Any assumptions about how results are being delivered -- beyond a pass/fail exit value -- are unnecessary and detrimental to adoption. Let the Dockerfile define the ONTEST VOLUME -- if, when and where it's needed.

But otherwise, this proposal is great. Maintaining separate Dockerfiles, and externally coordinating multiple builds, is a drag. Multistage build doesn't really solve the problem, since the test part gets thrown away, and the build fails (leaving you with nothing) if the tests fail. Providing a simple, standard way to define, build, and run a separate test image is so much better than existing alternatives.

**Comment by gaui** [comment]
@dmsteinberg Would be useful to be able to pipe the test results to stdout, but not sure how that would work because Dockerfile only has one CMD/ENTRYPOINT.
